bin/cron/pardot_sync

```diff
@@ -0,0 +1,419 @@
+
+    num_updated += 1
+
+    if record_count % UPDATE_BATCH_SIZE == 0
+      begin
+        conn.run update_batch
+      rescue => e
+        log "Error: #{e}. Cmd: #{update_batch}"
+        raise
+      end
+
+      # Discard the connection and get a new one after each batch. Currently, if you use multiple statements in a batch, the first call will succeed; the next call on
+      # the same connecton AFTER you do a multi-stament call will fail with a MySQL error that commands are out of order. The innertubes suggest that this may be a problem in mysql2 gem.
+      # Fortunately, it is fairly cheap to get a new connection for each batch (time to get new connection is negligible compared to batch time).
+      conn.disconnect
+      conn = mysql_multi_connection
+
+      update_batch = ""
+      if Time.now - time_last_output > 5
+        log "#{record_count} "
+        time_last_output = Time.now
+      end
+    end
+  end
+
+  conn.run update_batch unless update_batch.empty?
+  log "- #{record_count} source records, #{num_updated} with geo data, #{num_parse_errors} skipped due to invalid JSON, " + ((Time.now) - start).to_s + " seconds"
+end
+
+def update_grades_taught
+  start = Time.now
+  log "Updating grades taught from dashboard.users"
+  PEGASUS_DB.run "
+    UPDATE contacts_rollup,
+    (SELECT dashboard_user_id, group_concat(grade) as grades FROM
+	    (SELECT DISTINCT contacts_rollup.dashboard_user_id, sections.grade from contacts_rollup
+		    INNER JOIN #{DASHBOARD_DB_NAME}.sections AS sections on sections.user_id = contacts_rollup.dashboard_user_id
+		    WHERE contacts_rollup.dashboard_user_id is not null
+		    AND length(sections.grade) > 0
+		    ORDER by 1,2
+	    ) q
+	    GROUP BY dashboard_user_id
+    ) src
+    SET contacts_rollup.grades_taught = src.grades
+    WHERE contacts_rollup.id > 0 AND (contacts_rollup.dashboard_user_id = src.dashboard_user_id)"
+  log "   Completed in " + ((Time.now) - start).to_s + " seconds"
+end
+
+def update_ages_taught_for_age(age)
+  start = Time.now
+  log "Updating ages taught for age #{age}"
+
+  min_birthday = ""
+  max_birthday = ""
+  # for age = 21, include 21 and above
+  min_birthday = "students.birthday > DATE_ADD(NOW(),INTERVAL -#{age + 1} YEAR) AND" unless age >= 21
+  # for age = 4, include 4 and below
+  max_birthday = "students.birthday <= DATE_ADD(NOW(),INTERVAL -#{age} YEAR) AND" unless age <= 4
+
+  sql = "
+    UPDATE contacts_rollup, (
+    SELECT DISTINCT sections.user_id AS teacher_user_id
+    FROM #{DASHBOARD_DB_NAME}.users AS students
+      INNER JOIN #{DASHBOARD_DB_NAME}.followers AS followers
+        ON followers.student_user_id = students.id
+      INNER JOIN #{DASHBOARD_DB_NAME}.sections AS sections
+        ON sections.id = followers.section_id
+    WHERE
+      #{min_birthday}
+      #{max_birthday}
+      students.id != sections.user_id
```

Comment:
There are some teachers who have joined their own section who we need to filter out. This turned up because a bunch of teachers showed up as teaching ages like "6,7,21+", where the 21+ was themselves. I will clarify this in a comment.